proxy/handlers.go

```diff
@@ -147,3 +148,47 @@ func (p *Proxy) Impersonate(w http.ResponseWriter, r *http.Request) {
 
 	http.Redirect(w, r, dashboardURL, http.StatusFound)
 }
+
+// Verify checks a user's credentials for an arbitrary host. If the user is
+// properly authenticated and is authorized to access the supplied host,
+// a 200 http status code is returned. If the user is not authenticated, they
+// will be redirected to the authenticate service to sign in with their identity
+// provider. If the user is unauthorized, they will be given a 403 http status.
+func (p *Proxy) Verify(w http.ResponseWriter, r *http.Request) {
+	// retrieve the target destination hostname from the url path
+	hostname, ok := mux.Vars(r)["hostname"]
+	if !ok {
+		httputil.ErrorResponse(w, r, httputil.Error("no hostname set", http.StatusBadRequest, nil))
+		return
+	}
+	// attempt to retrieve the user session from the request context, validity
+	// of the identity session is asserted by the middleware chain
+	s, err := sessions.FromContext(r.Context())
```

Comment:
Implementation question - How are we getting access to the cookie on subsequent auth checks?  Does the third party proxy pass the cookies as part of the subrequest?

This depends on the proxy having an endpoint on the same domain that the requests hit the third party proxy with, correct?

Reply:
Great questions!

> How are we getting access to the cookie on subsequent auth checks?

Every request is checked for access and authorization. Every request will check and validate the session cookie. AFAI understand, this means every request to an endpoint results in another request to pomerium for validation. 

> > To perform authentication, NGINX makes an HTTP subrequest to an external server where the subrequest is verified.
https://docs.nginx.com/nginx/admin-guide/security-controls/configuring-subrequest-authentication/

> Does the third party proxy pass the cookies as part of the subrequest?

Yes, it will if pomerium sits on the same subdomain chain as the downstream application. Which paradoxically is why ...

> This depends on the proxy having an endpoint on the same domain that the requests hit the third party proxy with, correct?

For security reasons, I'd actually recommend _against_  having pomerium on the same domain because there's no explicit way to strip the session from downstream requests like we can do if we are the proxying entity.